rust-runtime/aws-smithy-wasm/src/wasi.rs

```diff
@@ -0,0 +1,350 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! WASI HTTP Adapter
+use aws_smithy_http::header::ParseError;
+use aws_smithy_runtime_api::{
+    client::{
+        http::{
+            HttpClient, HttpConnector, HttpConnectorFuture, HttpConnectorSettings,
+            SharedHttpClient, SharedHttpConnector,
+        },
+        orchestrator::HttpRequest,
+        result::ConnectorError,
+        runtime_components::RuntimeComponents,
+    },
+    http::Response,
+    shared::IntoShared,
+};
+use aws_smithy_types::body::SdkBody;
+use bytes::{Bytes, BytesMut};
+use wasi::http::{
+    outgoing_handler,
+    types::{self as wasi_http, OutgoingBody, RequestOptions},
+};
+
+/// Builder for [`WasiHttpClient`]. Currently empty, but allows for future
+/// config options to be added in a backwards compatible manner.
+#[derive(Default, Debug)]
+#[non_exhaustive]
+pub struct WasiHttpClientBuilder {}
+
+impl WasiHttpClientBuilder {
+    /// Creates a new builder.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Builds the [`WasiHttpClient`].
+    pub fn build(self) -> SharedHttpClient {
+        let client = WasiHttpClient {};
+        client.into_shared()
+    }
+}
+
+/// An HTTP client that can be used during instantiation of the client SDK in
+/// order to route the HTTP requests through the WebAssembly host. The host must
+/// support the WASI HTTP proposal as defined in the Preview 2 specification.
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct WasiHttpClient {}
+
+impl HttpClient for WasiHttpClient {
+    fn http_connector(
+        &self,
+        settings: &HttpConnectorSettings,
+        _components: &RuntimeComponents,
+    ) -> SharedHttpConnector {
+        let options = WasiRequestOptions::from(settings);
+        let connector = WasiHttpConnector { options };
+
+        connector.into_shared()
+    }
+}
+
+/// HTTP connector used in WASI environment
+#[derive(Debug, Clone)]
+pub struct WasiHttpConnector {
```

Comment:
Just one last change. It looks like this can be private, and since it's not possible to construct outside the crate, we should go ahead and do that to reduce the odds that we are constrained in the future due to backwards compatibility.
```suggestion
struct WasiHttpConnector {
```